### ISSUE TITLE
Implement vtables (for different calling conventions)

### ISSUE THREAD

**Issue by mhyee** [reactorlabs/rir#49]
Opening this for tracking/discussion.

**On Wed, Aug 30, 2017 at 8:31 AM, @o- wrote:**
>     The main idea for our next step was to support different calling
>     conventions (CC). As a first example we would like to have support for
>     value (instead of promise) CC.
> 
>     The idea is that a caller, who knows that the arguments are already
>     evaluated can call by value.
> 
>     For example "f(a+b)" or "f( (a) )"  (because the `(`-function is eager).
>     In this case we know the arg is a forced promise, so we might as well
>     pass it by value. Special care will be necessary to support reflection
>     in the callee, but that's another topic.
> 
>     As a first step we decided to introduce a layer of indirection in our
>     closures. Now the idea is that we have a vtable that can hold multiple
>     functions:
> 
>     closure
>        |---> formals
>        |---> env
>        |---> vtable
>                | ----> compiled_bc1
>                | ----> compiled_bc2
>                | ...
>                | ----> compiled_bcn
> 
>     the existing "call" bytecode will pick the first entry. And we'll add a
>     new bytecode that can pick other entries.

This issue is only for the implementation of the vtables. For reference, here's the plan for next steps (which will be tracked in a separate issue):

>     When that is ready the next step will be to extend the compiler to
>     support compiling under the assumption that all arguments are values
>     passed on the stack. and after that we'll have to find a solution for
>     reflection.
> 
>     even more advanced CC might come in the future. for example type
>     signature (akin julia dispatch), or compilation under assumptions (for
>     example "function compiled under the assumption that + is plus)".

**Comment by mhyee** [comment]
I've run into an issue.

Currently, our plan is to store the vtable in the closure's body:
```
closure
   |---> formals
   |---> env
   |---> body (actually a vtable)
           | ----> compiled_bc1
           | ----> compiled_bc2
           | ...
           | ----> compiled_bcn
```

(Each function has a pointer to the closure, so it can access the vtable it belongs to. Note that as we compile more versions of a function, the vtable may grow, so we have to allocate a larger one and update the body pointer.)

However, this breaks the API. `body` should return the function body, not the vtable:

```
> f <- function(x) x^5
> e <- expression(y <- x^2, return(y)) # or a list
> body(f) <- as.call(c(as.name("{"), e))
> f
function (x)
{
    y <- x^2
    return(y)
}
> f(8)
[1] 64
> body(f)
<external code: 0x3562038>
> substitute({ y <- x^2; return(y) })
{
    y <- x^2
    return(y)
}
>
> stopifnot(identical(body(f), substitute({ y <- x^2; return(y) })))
Error: identical(body(f), substitute({ .... is not TRUE
Execution halted
```

So setting the closure's body to be the vtable is not going to work. (I guess one solution is to modify `body` so it returns the first function in the vtable, but I'm not sure where to do that, and it might be related to the `BODY` macro which we do want to return the vtable.)

The obvious solution is to add a vtable field to the closure:
```
closure
   |---> formals
   |---> env
   |---> body (compiled_bc1)
   |---> vtable
           | ----> compiled_bc1
           | ----> compiled_bc2
           | ...
           | ----> compiled_bcn
```

But I worry that this would be an invasive change -- I would have to muck around in GNU R. (I'd also have to revert many of my changes, but oh well.)

Thoughts? @o- @petamaj @JanJecmen

**Comment by aviralg** [comment]
@mhyee Updating the C `body` implementation requires you to fiddle with the `do_body` function in `src/main/builtin.c`. This calls `BODY_EXPR` (different from `BODY` macro) which calls `R_ClosureExpr` which calls `bytecodeExpr` defined in `src/main/eval.c`. `bytecodeExpr` seems like a straightforward 6 line function. But as you mentioned, this means that we are modifying GNU R which may not be the solution you are looking for. 
Modifying `CLOSXP` (to add a `vtable` field) seems a bit more invasive than this.

**Comment by mhyee** [comment]
Thanks for the pointers! (I didn't know where `body` was defined and whether it was related to the `BODY` macro.)

I'll have to take a closer look tomorrow, but from your description, it sounds like that change would be far less invasive than modifying `CLOSXP`. Modifying GNU R is probably unavoidable. (But it does bother me slightly that "body" will mean different things in different contexts.)

**Comment by JanJecmen** [comment]
Looks to me as if the GNU R modification is already in place: `bytecodeExpr` from `eval.c` deals with `EXTERNALSXP` by calling RIR's handler, which gets installed by `initializeRuntime` (in `runtime.c`).

So the change could be made in the `rirExpr` function in `interp.c` (make it return some element of a vtable).

**Comment by mhyee** [comment]
Thanks @JanJecmen! That seems to have done the trick!

**Comment by mhyee** [comment]
Closed by #54.
